feat(common): add warnings re: image distortion

khempenius opened the pull request:
## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

AndrewKushnir commented on packages/common/src/directives/ng_optimized_image/ng_optimized_image.ts:
```diff
@@ -223,6 +224,10 @@ export class NgOptimizedImage implements OnInit, OnChanges, OnDestroy {
       assertRequiredNumberInput(this, this.width, 'width');
       assertRequiredNumberInput(this, this.height, 'height');
       assertValidLoadingInput(this);
+      const loadListener = this.renderer.listen(this.imgElement.nativeElement, 'load', () => {
+        loadListener();
```
I'd propose changing the name of this const to better reflect what it contains as a value.
```suggestion
      const removeListenerFn = this.renderer.listen(this.imgElement.nativeElement, 'load', () => {
        removeListenerFn();
```

AndrewKushnir commented on packages/common/src/directives/ng_optimized_image/ng_optimized_image.ts:
```diff
@@ -491,6 +496,64 @@ function assertRequiredNumberInput(dir: NgOptimizedImage, inputValue: unknown, i
+  const img = element.nativeElement;
+  const renderedWidth = parseFloat(img.clientWidth);
+  const renderedHeight = parseFloat(img.clientHeight);
+  const renderedAspectRatio = renderedWidth / renderedHeight;
```
Should we make sure the `renderedHeight` is not 0 (to avoid division by 0)? There are some other places in the code below that may be updated accordingly.

AndrewKushnir replied:
We can also take care of it later as a part of the "Warning or error if width or height is zero" task (from the tracking doc), but it'd be great to leave a TODO here in this case.

AndrewKushnir commented on packages/common/src/directives/ng_optimized_image/ng_optimized_image.ts:
```diff
@@ -491,6 +496,64 @@ function assertRequiredNumberInput(dir: NgOptimizedImage, inputValue: unknown, i
+  const stylingDistortion =
+      Math.abs(intrinsicAspectRatio - renderedAspectRatio) > aspectRatioTolerance;
+  if (inaccurateDimensions) {
+    console.log(Math.abs(suppliedAspectRatio - intrinsicAspectRatio));
+    console.log(Math.abs(intrinsicAspectRatio - renderedAspectRatio));
```
I guess these `console.log`s were here for debugging purposes and can be removed.

AndrewKushnir commented on packages/core/test/bundling/image-directive/e2e/image-distortion/image-distortion.e2e-spec.ts:
```diff
@@ -0,0 +1,85 @@
+
+    expect(logs.length).toEqual(5);
+
+    console.log(logs);
+
```
```suggestion
```

AndrewKushnir reviewed (commented):
@khempenius looks great, I've just left a few minor comments.

AndrewKushnir commented:
@khempenius it looks like there are some merge conflicts. Could you please rebase the PR?

Also, I noticed that the commit message lacks the description (i.e. there is just one line in the commit message), so the `lint` CI check would fail. Could you please also add some info to the commit message while rebasing?

khempenius committed:
feat(common): add warnings re: image distortion

Checks whether image is visually distorted. Also adds a check to verify that width and height are set to a non-zero number.

kara commented on packages/common/src/directives/ng_optimized_image/ng_optimized_image.ts:
```diff
@@ -413,6 +418,64 @@ function assertValidNumberInput(inputValue: unknown, inputName: string) {
+                  intrinsicAspectRatio}). ` +
+              `Rendered image size: ${renderedWidth}w x ${renderedHeight}h (aspect-ratio: ${
+                  renderedAspectRatio}). ` +
+              `To fix this, update the image styling.`);
```
Love that we're giving the diff in ratio and dimensions! Can we also add some of the advice in the comment above to the error message so people know how to update the styles? e.g. adding width or height: auto

kara commented on packages/common/src/directives/ng_optimized_image/ng_optimized_image.ts:
```diff
@@ -400,11 +404,12 @@ function assertNoPostInitInputChange(
-      (typeof inputValue === 'string' && /^\d+$/.test(inputValue.trim()));
-  if (!isValid) {
+// Verifies that a specified input is a number greater than 0.
+function assertGreatherThanZeroNumber(inputValue: unknown, inputName: string) {
```
Typo

```suggestion
function assertGreaterThanZeroNumber(inputValue: unknown, inputName: string) {
```

kara commented on packages/common/src/directives/ng_optimized_image/ng_optimized_image.ts:
```diff
@@ -146,6 +146,10 @@ export class NgOptimizedImage implements OnInit, OnChanges, OnDestroy {
       assertRequiredNumberInput(this, this.width, 'width');
       assertRequiredNumberInput(this, this.height, 'height');
       assertValidLoadingInput(this);
+      const removeListenerFn = this.renderer.listen(this.imgElement.nativeElement, 'load', () => {
```
Nit: move into separate function?

kara reviewed (approved):
This is fantastic! Excited for this feature to land :-) Some nits below

pkozlowski-opensource reviewed (approved):
LGTM

AndrewKushnir commented:
FYI there was a lint failure related to a missing semicolon in a test. I've pushed a change to add that semicolon and will proceed with merging the PR once the CI is "green". Thank you.
